UI testing

VasAbel opened the pull request.

VasAbel committed:
Tesztesetek és tesztkeret

A "Test cases" fájlban található a tervezett tesztek részletes folyamatleírása.
Elkészült egy keretprogram alapja, ami elindítja az alkalmazást, és a későbbiekben egymás után fogja futtatni a teszteket.

VasAbel committed:
1. teszt és framework javítás #1

Az első teszt (registration.py) elkészült és belekerült a frameworkbe. A framework kiegészült olyan megoldásokkal, amik a tesztek utáni adatbázis visszaállítását célozzák, hogy minden teszt tiszta állapotból induljon.
A script feltételezi, hogy telepítve van a dotnet-ef a %USERPROFILE%\.dotnet\tools útvonalon (ez az alapértelmezett telepítési könyvtár), illetve az npm.cmd a C:\Program Files\nodejs útvonalon (szintén alapértelmezett könyvtár)

VasAbel committed:
main javitasok, userlogin teszt #1

A mainre javításokat kellett eszközölni, hogy két teszt között az adatbázis újrainicializálódjon, és a weboldal a kezdőlapról induljon. Elkészült a 2. teszt userlogin néven, ami a későbbi tesztekhez is szükséges.

VasAbel committed:
3 új teszt

Elkészültek a productstocart, order, és adminlogin tesztek. A Test_cases.txt-ben elírás javítása. Dokumentáció elkezdése #1

VasAbel committed:
Kész #1

Minden teszt kész

19Makos99 commented on UItest/main.py:
```diff
@@ -0,0 +1,105 @@
+        driver.get("http://localhost:3000")
+        orderstatus.test_orderstatus(driver)
+        reset_database()                            #Change tests here, if you want to check a single one. You need to keep the test row and a reset_database row.
+        #userlogin.test_userlogin(driver)           #If multiple tests run after eachother, you also need to refresh the driver to ensure it didn't save anything from the previous test
+        #reset_database()
+        #driver = create_driver()
+        #driver.get("http://localhost:3000")
+        #ProductsToCart.test_cart(driver)
+        #reset_database()
+        #order.test_order(driver)
+        #reset_database()
```
Kikommentezett kód ne maradjon benne

19Makos99 commented on UItest/main.py:
```diff
@@ -0,0 +1,105 @@
+db_path = os.path.join(backend_path, 'WebShop.Api.csproj')
+dotnet_ef_path = os.path.join(os.environ['USERPROFILE'], '.dotnet', 'tools', 'dotnet-ef') #Change if dotnet-ef can be found elsewhere
+
+backend = None  # Global variable for the backend process
```
Globalt nem ajánlott használni

19Makos99 commented on doc/UI-Tesztek.md:
```diff
@@ -0,0 +1,34 @@
+
+A tesztek úgy készültek, hogy külön fájlokból egyenként is futtathatóak legyenek, ehhez azonban manuálisan kell indítani a backendet és frontendet, illetve a teszt után drop-olni és update-elni az adatbázist (ezek a main-ben meg vannak valósítva, így csak futtatni kell)
+
+A 6 teszt közül csak a regisztrációs és a megrendelés státusz teszt van külön meghívva a main-ben egymás után, mivel a megrendelés státusz futtatja a maradék 4 tesztet is, azokat felesleges lenne mégegyszer külön is futtatni. Amennyiben mégis szeretnénk valamelyik tesztet önállóan futtatni úgy, hogy ne kelljen manuálisan elindítanunk a tesztelt alkalmazást és visszaállítani az adatbázist, a main  ...
```
Ezt inkább indítási paraméterben kéne megadni

19Makos99 commented on doc/UI-Tesztek.md:
```diff
@@ -0,0 +1,34 @@
+Ekkor megnyílik egy Chrome böngésző, amelyen a script gombnyomásokat, szövegek beírását hajtja végig, így láthatjuk a tesztelt folyamatok vizuális lefutását, illetve a program a fontosabb vizsgálatoknál logolja az eredményeket.
+A futás végén a konzolon feltett _Terminate batch job?_ kérdésre "y"-vel válaszoljunk, ez állítja le a frontendet.
+
+_Szükséges függőségek_: 
```
A pip install selenium kimaradt
Ezt érdemes egy külön requirements.txt-be rakni

19Makos99 commented on doc/UI-Tesztek.md:
```diff
@@ -0,0 +1,34 @@
+
+A **dotnet ef**-nek megtalálhatónak kell lennie az alapértelmezett mappájában: _%USERPROFILE%\.dotnet\tools\dotnet-ef_. Ez a tesztelt alkalmazáshoz is kell elvileg, de ha mégsem itt található, ez a scriptben a jelzett helyen cserélendő a fájl elérési útjára.
+
+Az **npm**-nek is hasonlóan megtalálhatónak kell lennie a _C:\Program Files\nodejs\npm.cmd_ alapértelmezett útvonalon. Amennyiben mégsem itt van, szintén cserélendő a scriptben a jelzett helyen.
```
Ezt szerintem simán npm útvonalon is eléri, de nem biztos

19Makos99 commented on UItest/Test cases.txt:
```diff
@@ -0,0 +1,13 @@
+Tesztesetek folyamata:
+
+1. Regisztráció: Bejelentkezés -> E-mail: ujfelhasznalo@gmail.com, Jelszó: UjFelhasznalo123; elvárt kimenet: hiba -> Regisztráció -> Név: UjFelhasznalo, E-mail: ujfelhasznalo@gmail.com, Jelszó: UjFelhasznalo123; elvárt kimenet: bejelentkezett felület UjFelhasznalo néven -> Kijelentkezés -> Bejelentkezés -> E-mail: ujfelhasznalo@gmail.com, Jelszó: UjFelhasznalo123; elvárt kimenet: bejelentkezett f ...
+
+2. User bejelentkezés: Bejelentkezés -> E-mail: user@user.com, Jelszó: USER_user; elvárt kimenet: bejelentkezett felület user néven
+
+3. Termékek kosárba: (2) -> Termékek -> Sajt Add to cart -> Élelmiszer; elvárt kimenet: csak élelmiszerek felsorolva -> Sajt Add to cart -> Elektronika -> Ár 79000 max; elvárt kimenet: nincs nagyobb érték felsorolva -> Alaplap Add to cart -> Sport -> Ár 4000 és 95000 közt; elvárt kimenet: csak eközötti értékek felsorolva -> Roller Add to cart -> Kosár; elvárt kimenet: 2 sajt, 1 alaplap, 1 roller  ...
+
+4. Megrendelés: (3) -> Megrendel; elvárt kimenet: Sajt és roller felsorolva a termékeknél, Status: Unread -> Kosár; elvárt kimenet: üres kosár -> Megrendelések -> Számla; elvárt kimenet: pdf megnyitva
+
+5. Admin bejelentkezés: Bejelentkezés -> E-mail: admin@admin.com, Jelszó: ADMIN_admin; elvárt kimenet: bejelentkezett felület admin néven
+
+6. Megrendelés státusz: (4) -> Kijelentkezés -> (5) -> Megrendelések -> Számla Processing; elvárt kimenet: Status: Processing -> Kijelentkezés -> (2) -> Megrendelések; elvárt kimenet: Status: Processing
```
Ezt lehetne egy kicsit érthetőbben több szöveggel és kevesebb szimbólummal leírni

19Makos99 commented on UItest/ProductsToCart.py:
```diff
@@ -0,0 +1,162 @@
+#Used to check if given items are visible after the price interval has changed
+def check_item_visibility(driver, item_text, expected_visibility=True):
+    try:
+        item = driver.find_element(By.XPATH, f"//*[contains(text(), '{item_text}')]")
```
Nem használt változó

19Makos99 reviewed (changes requested):
A .idea és a __pycache__ ne kerüljön fel a repositoryba

VasAbel committed:
Jelzett problémák javítása #1

A .idea és pycache mappák kikerültek a repositoryból
A globális változó helyett osztályba szerveződött a backend változó és az azt kezelő függvények.
A teszt a futtatáskor megadott 1-7-ig terjedő paraméterrel választható ki, ahol 1-6-ig a dokumentációban található tesztet jelenti a szám, a 7 pedig az összes tesztet futtatja egymás után. Ezzel a kikommentezett kód is kikerült a fájlból.
A selenium és a többi függőség bekerült egy külön requirements.txt fájlba.
Az npm elérési útját meghagytam, mert nagyobb szabadságot hagy abban az esetben, ha egy directory-ban nem ismerné fel a parancsot a cli.
A Test cases.txt fájlban a teszt flow-kat bővebben leírtam.
A nem használt item változót kivettem a productstocart fájlból.

19Makos99 commented on UItest/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+
+npm: megtalálhatónak kell lennie a C:\Program Files\nodejs\npm.cmd alapértelmezett útvonalon. Amennyiben mégsem itt van, cserélendő a scriptben a jelzett helyen.
+
+selenium: pip install selenium segítségével telepítendő
```
Sorry nem részleteztem.
A requirements.txt fájl alatt a telepítendő pip csomagok és azok verzióját értettem, hogy ha frissítik az api-t, akkor ne törjön el.
Illetve nem kell a telepítendő csomagok listáját kézzel felrakni, csak a `pip install -r requirements.txt`-t kell kiadni.
Csak ennyit gondoltam a fájlba:
selenium=4.20.0

VasAbel committed:
requirements.txt

A requirements.txt használható a pip install kiadásához, illetve a kapcsolódó részek a dokumentációban átírásra kerültek.

19Makos99 reviewed (approved):
Oké, mindent rendben találtam.
Szép munka!
